orderer/common/multichannel/registrar.go

```diff
@@ -438,3 +443,66 @@ func (r *Registrar) RemoveChannel(channelID string, removeStorage bool) error {
 	//TODO
 	return errors.New("Not implemented yet")
 }
+
+// validateJoinBlock returns whether this block can be used as a join block for the channel participation API
+// as well as whether the orderer is a member of the consenter set.
+// A join block is a block of either a system channel containing a ConsortiumsConfig, or an application channel
+// containing an ApplicationConfig.
+func (r *Registrar) validateJoinBlock(channelID string, configBlock *cb.Block) (isAppChannel bool, isMember bool, err error) {
+	if !protoutil.IsConfigBlock(configBlock) {
+		return false, false, errors.New("block is not a config block")
+	}
+
+	envelope, err := protoutil.ExtractEnvelope(configBlock, 0)
+	if err != nil {
+		return false, false, err
+	}
+
+	bundle, err := channelconfig.NewBundleFromEnvelope(envelope, r.bccsp)
+	if err != nil {
+		return false, false, err
+	}
+
+	// Check channel id from join block matches
+	if bundle.ConfigtxValidator().ChannelID() != channelID {
+		return false, false, fmt.Errorf("config block channelID %s does not match passed channelID %s",
+			bundle.ConfigtxValidator().ChannelID(), channelID)
+	}
+
+	// Check channel type
+	_, isSystemChannel := bundle.ConsortiumsConfig()
+	if !isSystemChannel {
+		_, isAppChannel = bundle.ApplicationConfig()
+		if !isAppChannel {
+			return false, false, errors.New("invalid config: must have at least one of application or consortiums")
+		}
+	}
+
+	// Check if orderer is member of consenter set in join block
+	ordererConfig, ok := bundle.OrdererConfig()
+	if !ok {
+		return false, false, errors.New("channel config missing orderer configuration")
+	}
+
+	consensusType := ordererConfig.ConsensusType()
+	if consensusType != "etcdraft" {
+		return false, false, fmt.Errorf("join block consensus type is not configured to etcdraft, found type %s", consensusType)
+	}
+
+	// Retrieve etcdraft consenter so we can validate consensus metadata
+	consenter, ok := r.consenters["etcdraft"]
+	if !ok {
+		return false, false, errors.New("cannot find consenter of type etcdraft")
+	}
+
+	joinBlockValidator, ok := consenter.(consensus.JoinBlockValidator)
```

Comment:
Ideally we'll have an object which we can inject this thing into.

What we're doing here is always assuming we have etcdraft consenter, yet at the same time we are trying to be generic by using a JoinblockValidation.

Let's either be generic or not be generic, but this looks to me like something in between which is neither.

Reply:
By generic do you mean to have the kafka and solo consenters implement this and return error("join unsupported") or something? we can do that.

Reply:
I am saying the opposite - why have "JoinBlockValidator" for all consenters if it makes sense only to have it for cluster types?

Reply:
You said either or...
Anyway, we chose introspection in order to avoid changing the solo and kafka consenters. But the change is minimal really, 3 lines each. Having it in all would allow us to reject solo and kafka without even looking explicitly at the consensusType and spelling "etcdraft" in the registrar. One may argue it is cleaner. 

The `JoinBlockValidator` was "inspired" so to speak by the [MetadataValidator](https://github.com/tock-ibm/fabric/blob/73e5a537deeacc55dff3b352b478a4ae4e45a625/orderer/consensus/consensus.go#L32) interface that also targets raft only without binding to it directly... maybe its the wrong pattern...

Reply:
I said either or because I meant that either we are 100% generic or we are not, but what we have in this code is an attempt to be generic in the global structure but actually when you look deep enough you expect to have only a single type of consenter so we're not generic, hence delegating this consensus specific validation makes little sense because there is only a single implementation of it anyway.

What you're trying to do here is to just follow the existing philosophy of the orderer where everything in the multichannel invokes the consensus support interface, but i think the interface is really overly complex at this moment and is doing way too many stuff than it really should. 

Although it's convenient to extend this pattern i think we should avoid it here because there is really a much easier way to achieve what we want, and that is just abstract out what we need from the consensus validation checks (which i believe is a single function) and inject it into a validation object which the ParticipationAPI would delegate all validation of join channel to.

This has the benefit of reducing the responsibility of the registrar